Kernel/FileSystem/VFSRootContext.cpp

```diff
@@ -116,32 +116,225 @@ ErrorOr<NonnullRefPtr<VFSRootContext>> VFSRootContext::create_with_empty_ramfs()
+    });
+}
+
+void VFSRootContext::attach(Badge<Process>)
+{
+    m_details.with([](auto& details) {
+        details.attached_by_process.set();
+        details.attach_count++;
     });
 }
 
 bool VFSRootContext::mount_point_exists_at_custody(Custody& mount_point)
 {
-    return m_mounts.with([&](auto& mounts) -> bool {
-        return any_of(mounts, [&mount_point](auto const& existing_mount) {
+    return m_details.with([&](auto& details) -> bool {
+        return any_of(details.mounts, [&mount_point](auto const& existing_mount) {
             return existing_mount.host_custody() && VirtualFileSystem::check_matching_absolute_path_hierarchy(*existing_mount.host_custody(), mount_point);
         });
     });
 }
 
+ErrorOr<void> VFSRootContext::do_on_mount_for_host_custody(ValidateImmutableFlag validate_immutable_flag, Custody const& current_custody, Function<void(Mount&)> callback) const
+{
+    VERIFY(validate_immutable_flag == ValidateImmutableFlag::Yes || validate_immutable_flag == ValidateImmutableFlag::No);
+    return m_details.with([&](auto& details) -> ErrorOr<void> {
+        // NOTE: We either search for the root mount or for a mount that has a parent custody!
+        if (!current_custody.parent()) {
+            for (auto& mount : details.mounts) {
+                if (!mount.host_custody()) {
+                    if (validate_immutable_flag == ValidateImmutableFlag::Yes)
+                        TRY(VFSRootContext::validate_mount_not_immutable_while_being_used(details, mount));
+
+                    callback(mount);
+                    return {};
+                }
+            }
+            // NOTE: There must be a root mount entry, so fail if we don't find it.
+            VERIFY_NOT_REACHED();
+        } else {
+            for (auto& mount : details.mounts) {
+                if (mount.host_custody() && VirtualFileSystem::check_matching_absolute_path_hierarchy(*mount.host_custody(), current_custody)) {
+                    if (validate_immutable_flag == ValidateImmutableFlag::Yes)
+                        TRY(VFSRootContext::validate_mount_not_immutable_while_being_used(details, mount));
+
+                    callback(mount);
+                    return {};
+                }
+            }
+        }
+        return Error::from_errno(ENODEV);
+    });
+}
+
+ErrorOr<void> VFSRootContext::apply_to_mount_for_host_custody(Custody const& current_custody, Function<void(Mount&)> callback)
+{
+    return do_on_mount_for_host_custody(ValidateImmutableFlag::Yes, current_custody, move(callback));
+}
+
+ErrorOr<VFSRootContext::CurrentMountState> VFSRootContext::current_mount_state_for_host_custody(Custody const& current_custody) const
+{
+    RefPtr<FileSystem> guest_fs;
+    RefPtr<Inode> guest;
+    int mount_flags_for_child { 0 };
+    TRY(const_cast<VFSRootContext&>(*this).do_on_mount_for_host_custody(ValidateImmutableFlag::No, current_custody, [&guest, &guest_fs, &mount_flags_for_child](auto const& mount) {
```

Comment:
I don't think you need a `const_cast` here? `do_on_mount_for_host_custody` is already `const` (it can be because you made `m_details` mutable).
`m_details` being mutable is a bit iffy, but there isn't a clean solution for that I think. Otherwise we would have to keep the two very similar functions you had before (this might have been your reason for having them, sorry for not noticing)

Reply:
Well I guess we could use deducing this here?

Reply:
Ah we are only on GCC 13, so no deducing this sadly.

Reply:
Done.